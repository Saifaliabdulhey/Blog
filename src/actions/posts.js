import database from '../firebase/firebase'

// ADD POSTS
export const addPost = (post) => ({
  type: "ADD_POST",
  post
});
export const startAddPosts = (postData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      title = '',
      description = ''
    } = postData;

    const post = { description, title };

    return database.ref(`users/${uid}/posts`).push(post).then((ref) => {
      dispatch(addPost({
        id: ref.key,
        ...post
      }))
    })
  }
}


//REMOVE POST
export const removePost = ({ id } = {}) => ({
  type: 'REMOVE_POST',
  id
});

export const startRemovePost = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/posts/${id}`).remove().then(() => {
      dispatch(removePost({ id }));
    })
  }
}


//EDIT_POST
export const editPost = (id, updates) => ({
  type: 'EDIT_POST',
  id,
  updates
})

export const startEditpost = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/posts/${id}`)
      .update(updates)
      .then(() => dispatch(editPost(id, updates)))
  }
}

// SET_POSTS 
export const setPosts = (posts) => ({
  type: "SET_POSTS",
  posts
})
export const startSetPosts = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/posts`).once('value').then((snapshot) => {
      const posts = [];
      snapshot.forEach((childSnapshot) => {
        posts.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        })
      })
      dispatch(setPosts(posts));
    })
  }
}
