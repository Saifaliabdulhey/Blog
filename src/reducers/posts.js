const postsReducerDefaultState = [];
  
const postsReducer = (state = postsReducerDefaultState, action) => {
  switch (action.type) {
    case "ADD_POST":
      return [...state, action.post];
      case 'EDIT_POST':
        return state.map((post) => {
          if(post.id === action.if) {
            return {
              ...post, 
              ...action.updates
            };
          }else {
            return post;
          }
        })
      case 'REMOVE_POST':
        return state.filter(({ id }) => id !== action.id)
      case 'SET_POSTS':
        return action.posts;
    default:
      return state;
  }
};


export default postsReducer;