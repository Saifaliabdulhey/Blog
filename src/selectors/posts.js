const getVisiblePosts = (posts, text) => {
    return posts.filter((post) => {
        const textMatch = post.title
            .toLowerCase()
            .includes(text.toLowerCase());
        return textMatch
    })
}


export default getVisiblePosts;