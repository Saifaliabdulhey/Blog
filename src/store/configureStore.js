import {createStore, combineReducers, applyMiddleware, compose} from 'redux';
import thunk from 'redux-thunk';
import postsReducer from '../reducers/posts';
import authReducer from '../reducers/auth';

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
const exportStore = () => {
const store = createStore(
    combineReducers({
      posts: postsReducer,
      auth: authReducer
    }),
    composeEnhancers(applyMiddleware(thunk))
    // window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__()
  );
  return store;
}

export default exportStore;