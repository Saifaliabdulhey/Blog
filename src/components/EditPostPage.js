import React, { Component, setState } from 'react';
import { connect } from 'react-redux';
import PostForm from './PostForm';
import { startRemovePost, startEditpost } from '../actions/posts';
import { Link } from 'react-router-dom'

export class EditPostPage extends Component {
    onSubmit = (post) => {
        this.props.startEditpost(this.props.post.id, post);
        this.props.history.push('/home');
    }
    onRemove = () => {
        this.props.startRemovePost({ id: this.props.post.id });
        this.props.history.push('/home');
    }

    render() {
        let path = window.location.origin;
        const id = this.props.match.params.id;
        return (
            <div>
                <div className="container">
                    <h1>Edit Post</h1>
                </div>
                <div className="container">
                    <PostForm
                        post={this.props.post}
                        onSubmit={this.onSubmit} />
                    <div className="container btn-container">
                        <button  onClick={this.onRemove}>Remove Post</button>
                    </div>
                    <Link className="read_link" to={`/read/${id}`}>
                    <h4><span>Post is Readable at:</span>{`${path}/read/${id}`}</h4>
                    </Link>
                    
                </div>
            </div>
        )
    }
}

const mapStateToProps = (state, props) => ({
    post: state.posts.find((post) => post.id === props.match.params.id)
})

const mapDispatchToProps = (dispatch, props) => ({
    startEditpost: (id, post) => dispatch(startEditpost(id, post)),
    startRemovePost: (data) => dispatch(startRemovePost(data))
})

export default connect(mapStateToProps, mapDispatchToProps)(EditPostPage);