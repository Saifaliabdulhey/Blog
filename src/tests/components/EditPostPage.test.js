import React from 'react';
import { shallow } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from '../../store/configureStore';
import posts from '../fixture/posts';
import EditPostPage  from '../../components/AddPostPage';


const store = configureStore();

let startEditPost, startRemovePost, history, wrapper;

beforeEach(() => {
    startEditPost = jest.fn();
    startRemovePost = jest.fn();
    history = { push: jest.fn() };
    wrapper = shallow(
        <Provider store={store}>
        <EditPostPage
            startEditPost={startEditPost}
            startRemovePost={startRemovePost}
            history={history}
            post={posts[1]} />
            </Provider>
    )
})

test('should render EditPostPage', () => {
    expect(wrapper).toMatchSnapshot();
})

// test('should handle startEditPost', () => {
//     wrapper.find('PostForm').prop('onSubmit')(posts[2]);
//     expect(history.push).toHaveBeenLastCalledWith('/');
//     expect(startEditPost).toHaveBeenLastCalledWith(posts[2].id, posts[2]);
// })

// test('should handle startRemovePost', () => {
//     wrapper.find('button').simulate('click');
//     expect(history.push).toHaveBeenLastCalledWith('/');
//     expect(startRemovePost).toHaveBeenLastCalledWith({
//         id: posts[2].id
//     })
// })

