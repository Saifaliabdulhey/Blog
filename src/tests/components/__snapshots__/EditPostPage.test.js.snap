// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditPostPage 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <Context.Provider
    value={
      Object {
        "store": Object {
          "@@observable": [Function],
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "subscription": Object {
          "addNestedSub": [Function],
          "getListeners": [Function],
          "handleChangeWrapper": [Function],
          "isSubscribed": [Function],
          "notifyNestedSubs": [Function],
          "trySubscribe": [Function],
          "tryUnsubscribe": [Function],
        },
      }
    }
  >
    <Memo(Connect(AddPostPage))
      history={
        Object {
          "push": [MockFunction],
        }
      }
      post={
        Object {
          "description": "Rent",
          "id": "2",
          "title": "new",
        }
      }
      startEditPost={[MockFunction]}
      startRemovePost={[MockFunction]}
    />
  </Context.Provider>,
  "nodes": Array [
    <Context.Provider
      value={
        Object {
          "store": Object {
            "@@observable": [Function],
            "dispatch": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          },
          "subscription": Object {
            "addNestedSub": [Function],
            "getListeners": [Function],
            "handleChangeWrapper": [Function],
            "isSubscribed": [Function],
            "notifyNestedSubs": [Function],
            "trySubscribe": [Function],
            "tryUnsubscribe": [Function],
          },
        }
      }
    >
      <Memo(Connect(AddPostPage))
        history={
          Object {
            "push": [MockFunction],
          }
        }
        post={
          Object {
            "description": "Rent",
            "id": "2",
            "title": "new",
          }
        }
        startEditPost={[MockFunction]}
        startRemovePost={[MockFunction]}
      />
    </Context.Provider>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_context": Object {},
    "_didScheduleRenderPhaseUpdate": false,
    "_dispatcher": Object {
      "readContext": [Function],
      "useCallback": [Function],
      "useContext": [Function],
      "useDebugValue": [Function],
      "useDeferredValue": [Function],
      "useEffect": [Function],
      "useId": [Function],
      "useImperativeHandle": [Function],
      "useInsertionEffect": [Function],
      "useLayoutEffect": [Function],
      "useMemo": [Function],
      "useReducer": [Function],
      "useRef": [Function],
      "useResponder": [Function],
      "useState": [Function],
      "useSyncExternalStore": [Function],
      "useTransition": [Function],
    },
    "_element": <Provider
      store={
        Object {
          "@@observable": [Function],
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        }
      }
    >
      <Memo(Connect(AddPostPage))
        history={
          Object {
            "push": [MockFunction],
          }
        }
        post={
          Object {
            "description": "Rent",
            "id": "2",
            "title": "new",
          }
        }
        startEditPost={[MockFunction]}
        startRemovePost={[MockFunction]}
      />
    </Provider>,
    "_firstWorkInProgressHook": Object {
      "memoizedState": Array [
        Object {
          "store": Object {
            "@@observable": [Function],
            "dispatch": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          },
          "subscription": Object {
            "addNestedSub": [Function],
            "getListeners": [Function],
            "handleChangeWrapper": [Function],
            "isSubscribed": [Function],
            "notifyNestedSubs": [Function],
            "trySubscribe": [Function],
            "tryUnsubscribe": [Function],
          },
        },
        Array [
          Object {
            "@@observable": [Function],
            "dispatch": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          },
        ],
      ],
      "next": Object {
        "memoizedState": Array [
          Object {
            "auth": Object {},
            "filters": Object {
              "text": "",
            },
            "posts": Array [],
          },
          Array [
            Object {
              "@@observable": [Function],
              "dispatch": [Function],
              "getState": [Function],
              "replaceReducer": [Function],
              "subscribe": [Function],
            },
          ],
        ],
        "next": null,
        "queue": null,
      },
      "queue": null,
    },
    "_forcedUpdate": false,
    "_idCounter": 0,
    "_instance": null,
    "_isReRender": false,
    "_newState": null,
    "_numberOfReRenders": 0,
    "_renderPhaseUpdates": null,
    "_rendered": <Context.Provider
      value={
        Object {
          "store": Object {
            "@@observable": [Function],
            "dispatch": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          },
          "subscription": Object {
            "addNestedSub": [Function],
            "getListeners": [Function],
            "handleChangeWrapper": [Function],
            "isSubscribed": [Function],
            "notifyNestedSubs": [Function],
            "trySubscribe": [Function],
            "tryUnsubscribe": [Function],
          },
        }
      }
    >
      <Memo(Connect(AddPostPage))
        history={
          Object {
            "push": [MockFunction],
          }
        }
        post={
          Object {
            "description": "Rent",
            "id": "2",
            "title": "new",
          }
        }
        startEditPost={[MockFunction]}
        startRemovePost={[MockFunction]}
      />
    </Context.Provider>,
    "_rendering": false,
    "_updater": Updater {
      "_callbacks": Array [],
      "_renderer": [Circular],
    },
    "_workInProgressHook": null,
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Provider
    store={
      Object {
        "@@observable": [Function],
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  >
    <Memo(Connect(AddPostPage))
      history={
        Object {
          "push": [MockFunction],
        }
      }
      post={
        Object {
          "description": "Rent",
          "id": "2",
          "title": "new",
        }
      }
      startEditPost={[MockFunction]}
      startRemovePost={[MockFunction]}
    />
  </Provider>,
}
`;
